src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1041,102 +1041,190 @@ private bool IsValidFixedVariableInitializer(TypeSymbol declType, SourceLocalSym
+                return false;
+            }
 
-                    // CONSIDER: this is a very confusing error message, but it's what Dev10 reports.
-                    Error(diagnostics, ErrorCode.ERR_FixedNotNeeded, initializerSyntax);
-                }
+            TypeSymbol elementType;
+            bool hasErrors = false;
+            MethodSymbol getPinnableMethod = null;
+
+            switch (initializerOpt.Kind)
+            {
+                case BoundKind.AddressOfOperator:
+                    elementType = ((BoundAddressOfOperator)initializerOpt).Operand.Type;
+                    break;
+
+                case BoundKind.FieldAccess:
+                    var fa = (BoundFieldAccess)initializerOpt;
+                    if (fa.FieldSymbol.IsFixed)
+                    {
+                        elementType = ((PointerTypeSymbol)fa.Type).PointedAtType;
+                        break;
+                    }
+
+                    goto default;
+
+                default:
+                    //  fixed (T* variable = <expr>) ...
+
+                    // check for arrays
+                    if (initializerType.IsArray())
+                    {
+                        // See ExpressionBinder::BindPtrToArray (though most of that functionality is now in LocalRewriter).
+                        elementType = ((ArrayTypeSymbol)initializerType).ElementType;
+                        break;
+                    }
+
+                    // check for a special ref-returning method
+                    var additionalDiagnostics = DiagnosticBag.GetInstance();
+                    getPinnableMethod = GetDangerousGetPinnableReferenceMethodOpt(initializerOpt, additionalDiagnostics);
+
+                    // check for String
+                    // NOTE: We will allow DangerousGetPinnableReferenceMethod to take precendence, but only if it is a member of System.String
+                    if (initializerType.SpecialType == SpecialType.System_String &&
+                        ((object)getPinnableMethod == null || getPinnableMethod.ContainingType.SpecialType != SpecialType.System_String))
+                    {
+                        elementType = this.GetSpecialType(SpecialType.System_Char, diagnostics, initializerSyntax);
+                        additionalDiagnostics.Free();
+                        break;
+                    }
+
+                    if ((object)getPinnableMethod != null)
+                    {
+                        elementType = getPinnableMethod.ReturnType;
+                        CheckFeatureAvailability(initializerOpt.Syntax, MessageID.IDS_FeatureExtensibleFixedStatement, diagnostics);
+                        additionalDiagnostics.Free();
+                        break;
+                    }
+                    else
+                    {
+                        // if feature was enabled but somethng went wrong with the method, report that
```

Comment:
I think we should report that the method was wrong regardless of the feature being enabled.
The errors should provide you a path (instructions) to enabling the feature and putting an appropriate method in place.

Reply:
Initially I did not consider that the diagnostics that we could not find a helper method is interesting at all. It was very noisy. 
We would get those errors anywhere we try to pin something that we cannot pin (since nearly everything now can be pinned via an extension method, if such available).

Then I figured that there is one situation when you want those errors - when you are designing the actual helper, so I added some heuristics to detect that.
My criteria are - the feature is enabled and the method is "almost there, but not quite right". 
Then I believe that someone is developing the helper and would want to see the extra diagnostics.

---
In reply to: [169494680](https://github.com/dotnet/roslyn/pull/24469#discussion_r169494680) [](ancestors = 169494680,169487046)

Reply:
OK. That works